File: GASEngineSource/Engine/CameraControllerA/Manipulator.js
GASEngine.Manipulator = function(type)
{
    this._controllerList = {};
    this._cameraWorldMatrix = new GASEngine.Matrix4();
    this._camera = undefined;
    this._type = type;
    this._controllerType = null;
};

GASEngine.Manipulator.ORBIT = 0;
GASEngine.Manipulator.FPS = 1;
GASEngine.Manipulator.AUTOPILOT = 2;

GASEngine.Manipulator.prototype =
{
    setCamera: function(c)
    {
        this._camera = c;
    },

    getCamera: function()
    {
        return this._camera;
    },

    setControllerType: function(type)
    {
        this._controllerType = type;
    },

    getControllerType: function()
    {
        return this._controllerType;
    },

    getType: function()
    {
        return this._type;
    },

    getHomeDistance: function(sceneRadius)
    {
        var dist = sceneRadius;
        if(this._camera)
        {
            var top = Math.tan(GASEngine.degToRad(this._camera.fov * 0.5)) * this._camera.near;
            var bottom = -top;
            var left = this._camera.aspect * bottom;
            var right = this._camera.aspect * top;

            var near = this._camera.near;

            var vertical2 = Math.abs(right - left) / near / 2;
            var horizontal2 = Math.abs(top - bottom) / near / 2;
            dist /= Math.sin(Math.atan2(horizontal2 < vertical2 ? horizontal2 : vertical2, 1));
        }
        else
        {
            dist *= 1.5;
        }

        return dist;
    },

    update: function()
    {
    },

    getCameraWorldMatrix: function()
    {
        return this._cameraWorldMatrix;
    },

    getCurrentController: function()
    {
        if(this._controllerType === null)
            return null;

        return this._controllerList[this._controllerType];
    }
};

File: GASEngineSource_OpenGL/src/manipulator/manipulator.cpp
#include "manipulator/manipulator.h"
#include <math.h>
#include <algorithm>
#include "ecs/component/camera_component.h"
#include "glm/gtx/euler_angles.hpp"

float Manipulator::GetHomeDistance(float scene_radius) {
    if (camera != nullptr && camera->camera_type == kPerspective) {
        float top = std::tan(camera->field_of_view * 0.5f) * camera->near;
        float bottom = -top;
        float left = bottom * camera->aspect;
        float right = top * camera->aspect;
        float vertical2 = std::abs(right - left) / camera->near / 2.0f;
        float horizontal2 = std::abs(top - bottom) / camera->near / 2.0f;
        float divisor = std::sin(std::atan2(std::min(horizontal2, vertical2), 1.0f));
        return scene_radius / divisor;
    } else {
        return 1.5 * scene_radius;
    }
}

float Manipulator::GetSpeedFactor() const { return std::max(distance, min_speed); }

void Manipulator::ComputeEyePosition() {
    this->eye = target + rotation * glm::vec3(0.0f, 0.0f, std::max(distance, min_distance));
}

void Manipulator::SetTarget(const glm::vec3& target) { this->target = target; }